### PR TITLE
CDAP-1878 Add Kafka realtime source

### PR DESCRIPTION
**Pull request by henry-cask**
The PR contains first drop supporting Kafka as realtime source:
-) The Kafka consumer are borrowed from cdap-kafka-pack/cdap-kafka-flow project leveraging Kafka Simple Consumer. The logic is modified and simplify to fit the scenario of realtime source which is less complicated that CDAP Flow and Flowlets.
-) The Kafka source only support latest Kafka version supported by CDAP which is 0.8.x (as long as Kafka does not have API change within minor releases). 
-) The Kafka realtime source also leverages library from Twill to start and stop ZK server and client and manage the Topic partition and offset.
-) Add unit test to test single partition with single broker list for send and receive Kafka messages. 

TODO: 
-) Will add more tests such as to use ZK as connection broker

### CODE REVIEW

**Comment by henry-cask** [comment]
CI: 
http://builds.cask.co/browse/CDAP-DT37-4

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/DefaultKafkaConfigurer.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/DefaultKafkaConfigurer.java]: 2015

**Comment by henry-cask** [comment]
Latest build passed:
http://builds.cask.co/browse/CDAP-DT37-6

**Comment by henry-cask** [comment]
If no more comment will merge this end of day today

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+    return getClass().getSimpleName();
+  }
+
+  public Map<String, byte[]> getCopyOffsetStore() {
+    return Maps.newHashMap(offsetStore);
+  }
+
+  /**
+   *  Set the internal offset store for the Topic partition.
+   */
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: I would call this saveState()

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+
+  /**
+   *
+   * @return the name of this Kafka realtime source. Default would be the simple class name.
+   */
+  public String getName() {
+    return getClass().getSimpleName();
+  }
+
+  public Map<String, byte[]> getCopyOffsetStore() {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: I would call this getSavedState()

Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: Agree, will make the change.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java (hunk 2 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: Agree.

**Comment by henry-cask** [comment]
CC @nitinmotgi

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+public class KafkaSource extends RealtimeSource<ByteBuffer> {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaSource.class);
+
+  private static final String CDAP_KAFKA_SOURCE_NAME = "Kafka Realtime Source";
+
+  private KafkaSimpleApiConsumer kafkaConsumer;
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(CDAP_KAFKA_SOURCE_NAME);
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: should remove this, so that the ``Source`` can be identified by its class name, instead of a custom string.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    super.initialize(context);
+
+    kafkaConsumer = new Kafka08SimpleApiConsumer();
+    kafkaConsumer.initialize(context);
+  }
+
+  @Nullable
+  @Override
+  @SuppressWarnings("unchecked")
+  public SourceState poll(Emitter<ByteBuffer> writer, SourceState currentState) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: Should use ``Emitter<StructuredRecord>`` here, since we are following the convention that all sources emit ``StructuredRecord``s, while sinks and transforms consume ``StructuredRecord``s.

You will have to write a class to transform the ``ByteBuffer`` into a ``StructuredRecord``.

The ``StructuredRecord`` can possibly have two fields - 'headers' and 'body', where the body can contain the ``ByteBuffer`` to emit. ``StreamBatchSource`` should have an example.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSourceTest.java [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+import java.util.concurrent.TimeUnit;
+
+/**
+ * <p>
+ *  Unit test for {@link KafkaSource} ETL realtime source class.
+ *
+ *  The base code is borrowed from {@code cdap-kafka-pack} project.
+ * </p>
+ */
+public class KafkaSourceTest {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSourceTest.java]: Would be good to create a follow-up task to write an end-to-end integration test - On the lines of ``ETLWorkerTest``, that currently uses ``TestSource`` for lack of a better source. 

However, lets do that later, since that test will probably get revamped a little after plugin system integration.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(Kafka08SimpleApiConsumer.class);
+
+  private ZKClientService zkClient;
+  private BrokerService brokerService;
+  private Cache<TopicPartition, SimpleConsumer> kafkaConsumers;
+
+  @Override
+  protected void configureKafka(KafkaConfigurer configurer) {
+    Map<String, String> runtimeArgs = getContext().getRuntimeArguments();
+    if (runtimeArgs.containsKey("kafka.zookeeper")) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: Not too familiar with Kafka, but what happens when neither of these keys are present? I see that both these params are not set?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(CDAP_KAFKA_SOURCE_NAME);
+    configurer.setDescription(CDAP_KAFKA_SOURCE_NAME);
+    configurer.addProperty(new Property("kafka.zookeeper ", "The connect string location of Kafka Zookeeper", false));
+    configurer.addProperty(new Property("kafka.brokers", "Comma separate list of Kafka brokers", false));
+    configurer.addProperty(new Property("kafka.partitions", "Number of partitions" , true));
+    configurer.addProperty(new Property("kafka.topic", "Topic of the messages", true));
+    configurer.addProperty(new Property("kafka.default.offset", "The default offset for the partition", false));
+  }
+
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: Should probably override the ``configurePipeline()`` method to validate the properties defined above. Things like some of them have to be ints, atleast one of two have to be present, etc. Also, setting true in the ``Property`` object does not ensure that the property exists, so should add ``Preconditions`` for them in the ``configurePipeline()`` method

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 5 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: The original design is to emit ByteBuffer for Kafka source and it is consistent with what Kafka message deliver from the Consumer API.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+    String kafkaTopic = runtimeArgs.get("kafka.topic");
+    for (int i = 0; i < partitions; i++) {
+      if ((i % instances) == instanceId) {
+        configurer.addTopicPartition(kafkaTopic, i);
+      }
+    }
+  }
+
+  @Override
+  protected Iterator<KafkaMessage<Long>> readMessages(KafkaConsumerInfo<Long> consumerInfo) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: ``consumerInfo`` can never be null, right? Otherwise add a ``Preconditions`` check

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+   * Creates a Kafka {@link BrokerService}.
+   */
+  private BrokerService createBrokerService(ZKClient zkClient) throws Exception {
+    Class<? extends BrokerService> brokerClass = loadBrokerServiceClass(BrokerService.class.getClassLoader());
+    Constructor<? extends BrokerService> constructor = brokerClass.getDeclaredConstructor(ZKClient.class);
+    constructor.setAccessible(true);
+    return constructor.newInstance(zkClient);
+  }
+
+
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: nit: extra newline

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java (hunk 7 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: In the initialize method this will be checked and we will throw IllegalStateException if none is being set.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 8 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: The configurePipeline is new for realtime source context so will add validation here.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSourceTest.java (hunk 6 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSourceTest.java]: Agree but that should not hold this PR.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+  /**
+   * Loads the class that implements {@link BrokerService}.
+   */
+  @SuppressWarnings("unchecked")
+  private <T> Class<T> loadBrokerServiceClass(ClassLoader classLoader) throws ClassNotFoundException {
+    // TODO: Hacky way to construct ZKBrokerService from Twill as it's a protected class
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: if there is no other way, perhaps create a Twill jira?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 4 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: Will use class name instead.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+      return 0L;
+    }
+    return response.offsets(topic, partition)[0];
+  }
+
+  /**
+   * Returns a {@link SimpleConsumer} for the given consumer info or {@code null} if no leader broker is currently
+   * available.
+   */
+  private SimpleConsumer getConsumer(KafkaConsumerInfo<Long> consumerInfo) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: This method should be marked ``@Nullable``, since it can return a null.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+    // Need Twill update to resolve this hack
+    return (Class<T>) classLoader.loadClass("org.apache.twill.internal.kafka.client.ZKBrokerService");
+  }
+
+  /**
+   * Gets the address of the leader broker for the given topic and partition
+   *
+   * @return the address for the leader broker or {@code null} if no leader is currently available.
+   */
+  private InetSocketAddress getLeader(String topic, int partition) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: ``@Nullable`` as well

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+    // Otherwise use broker list to discover leader
+    return findLeader(getKafkaConfig().getBrokers(), topic, partition);
+  }
+
+  /**
+   * Finds the leader broker address for the given topic partition.
+   *
+   * @return the address for the leader broker or {@code null} if no leader is currently available.
+   */
+  private InetSocketAddress findLeader(String brokers, String topic, int partition) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: ``@Nullable``

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 4 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: But why would that matter? Name could just be any String I believe?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+
+      keyDecoder = createKeyDecoder(typeArgs[0]);
+      payloadDecoder = createPayloadDecoder(typeArgs[1]);
+    }
+
+    // Configure kafka
+    DefaultKafkaConfigurer kafkaConfigurer = new DefaultKafkaConfigurer();
+    configureKafka(kafkaConfigurer);
+
+    if (kafkaConfigurer.getZookeeper() == null && kafkaConfigurer.getBrokers() == null) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: ah! looks like you catch that case I mentioned for validation earlier here. Hmmm, I think you can do this much earlier, in the ``configurePipeline()`` method, so this won't be needed.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 4 above)
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: Yes, it can be any string. However, its just one less string to remember while writing tests. If you just skip this method call, then it will default to the class name.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+  /**
+   * Overrides this method if interested in the raw Kafka message.
+   *
+   * @param message The message fetched from Kafka.
+   * @param emitter The emitter to push the message
+   */
+  protected void processMessage(KafkaMessage<OFFSET> message, Emitter<ByteBuffer> emitter) {
+    // Should only receive messages from partitions that it can process
+    int partition = message.getTopicPartition().getPartition();
+    if ((partition % getContext().getInstanceCount()) != getContext().getInstanceId()) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: Can use ``Preconditions.checkArgument`` instead - just convention

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+  /**
+   * Creates a decoder that convert the input {@link ByteBuffer} into UTF-8 String. The input {@link ByteBuffer}
+   * will not be consumed after the call.
+   */
+  private Function<ByteBuffer, String> createStringDecoder() {
+    return new Function<ByteBuffer, String>() {
+      @Override
+      public String apply(ByteBuffer input) {
+        input.mark();
+        String result = Charsets.UTF_8.decode(input).toString();
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: You can use ``Bytes.toString`` here instead of doing the decoding on your own.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+  /**
+   *  Set the internal state for the Topic partition.
+   */
+  public void saveState(SourceState currentState) {
+    // Update the internal state
+    offsetStore = currentState.getState();
+  }
+
+  /**
+   * WIll be called by external source to start poll the Kafka messages one at the time.
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: nit: lower-case 'i'

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+   */
+  protected abstract long getDefaultOffset(TopicPartition topicPartition);
+
+  /**
+   * Override this method to provide custom decoding of a message key.
+   *
+   * @param buffer The bytes representing the key in the Kafka message
+   * @return The decoded key
+   */
+  protected KEY decodeKey(ByteBuffer buffer) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: ``@Nullable``?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 20]
```diff
@@ -0,0 +1,415 @@
...
+    return (keyDecoder != null) ? keyDecoder.apply(buffer) : null;
+  }
+
+  /**
+   * Override this method to provide custom decoding of a message payload.
+   *
+   * @param buffer The bytes representing the payload in the Kafka message
+   * @return The decoded payload
+   */
+  protected PAYLOAD decodePayload(ByteBuffer buffer) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: ``@Nullable`` too?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java [hunk 21]
```diff
@@ -0,0 +1,415 @@
...
+
+  /**
+   * Creates a decoder that reads {@link ByteBuffer} content and return it as {@code byte[]}. The
+   * input {@link ByteBuffer} will not be consumed after the call.
+   */
+  private Function<ByteBuffer, byte[]> createBytesDecoder() {
+    return new Function<ByteBuffer, byte[]>() {
+      @Override
+      public byte[] apply(ByteBuffer input) {
+        byte[] bytes = new byte[input.remaining()];
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: You can use ``Bytes.toBytes(ByteBuffer)`` here

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaMessage.java [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.twill.kafka.client.TopicPartition;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Represents a Kafka message.
+ *
+ * @param <OFFSET> Type of message offset
+ */
+public class KafkaMessage<OFFSET> {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaMessage.java]: should this be ``final``

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaConfig.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.templates.etl.realtime.kafka;
+
+/**
+ * Contains information about Kafka cluster as configured by user.
+ */
+public class KafkaConfig {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaConfig.java]: should be ``final``?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaConfig.java [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+  public KafkaConfig(String zookeeper, String brokers) {
+    this.zookeeper = zookeeper;
+    this.brokers = brokers;
+  }
+
+  /**
+   * Returns the ZooKeeper connection string as set through {@link KafkaConsumerConfigurer#setZooKeeper(String)}
+   * or {@code null}.
+   */
+  public String getZookeeper() {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaConfig.java]: ``@Nullable``

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaConfig.java [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+    return zookeeper;
+  }
+
+  /**
+   * Returns brokers information as set through {@link KafkaConsumerConfigurer#setBrokers(String)} or {@code null}.
+   */
+  public String getBrokers() {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaConfig.java]: ``@Nullable`` too

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 5 above)
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: Yeah, however, this would mean that to apply any built-in transform, or to use any built-in sink within the ETL Adapters, we would have to add a ``Transform`` to transform the record to a ``StructuredRecord`` before you can do anything else.

I think this PR is already large, so if you'd like, we can make this change in a follow-up PR. However, I think we should do it.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSourceTest.java (hunk 6 above)
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSourceTest.java]: Yes. not for this PR.

**Comment by bdmogal** [comment]
Done with 1 pass. Mostly minor comments other than the one about emitting ``StructuredRecord``

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java (hunk 11 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: I did and have sent patch: https://github.com/apache/incubator-twill/pull/31

**Comment by bdmogal** [comment]
Also, merge conflicts

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java (hunk 12 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: Will do

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java (hunk 15 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: I'll keep them both for now to avoid much changes.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java (hunk 21 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java]: I think this is a bit different bc we are trying to read the cur position to end of buffer. Will keep it as is for now.

**Comment by henry-cask** [comment]
Send new build:
http://builds.cask.co/browse/CDAP-DT37-14

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java (hunk 11 above)
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: :+1:

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java (hunk 9 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/Kafka08SimpleApiConsumer.java]: Should be ok because the method is called with iteration to consumer info collections that the value is created with new KafkaConsumerInfo call.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSourceTest.java (hunk 6 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSourceTest.java]: Ah cool, thanks!

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 5 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: Agree, let's do it in separate PR then.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+  public static final String KAFKA_BROKERS = "kafka.brokers";
+  public static final String KAFKA_DEFAULT_OFFSET = "kafka.default.offset";
+
+  private static final String CDAP_KAFKA_SOURCE_NAME = "Kafka Realtime Source";
+
+  private KafkaSimpleApiConsumer kafkaConsumer;
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(getClass().getSimpleName());
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: this is redundant

Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: I mean it is done by default

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+    configurer.setDescription(CDAP_KAFKA_SOURCE_NAME);
+    configurer.addProperty(new Property(KAFKA_ZOOKEEPER, "The connect string location of Kafka Zookeeper", false));
+    configurer.addProperty(new Property(KAFKA_BROKERS, "Comma separate list of Kafka brokers", false));
+    configurer.addProperty(new Property(KAFKA_PARTITIONS, "Number of partitions" , true));
+    configurer.addProperty(new Property(KAFKA_TOPIC, "Topic of the messages", true));
+    configurer.addProperty(new Property(KAFKA_DEFAULT_OFFSET, "The default offset for the partition", false));
+  }
+
+  @Override
+  public void  configurePipeline(ETLStage stageConfig, PipelineConfigurer pipelineConfigurer) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: nit: extra space

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 26 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: Want to keep it now so I know explicitly the name of the source for now.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java (hunk 27 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/KafkaSource.java]: You are good, not sure how you catch it =)

**Comment by bdmogal** [comment]
Code-wise, (and also adapter business logic wise) this LGTM. I'm not too familiar with Kafka, so we may need another pass at this by someone who is.

Also, we need a follow-up to add a record transformer in the source, to emit ``StructuredRecord`` instead of the current ``ByteBuffer``.

**Comment by sreevatsanraman** [comment]
I am merging this. We will address any comments in follow-up PR>
